File: src/cpp_basic_nodes/src/launch1_subscriber.cpp
/*
 * This is a simple subscriber that is a part of `launch1`
 * 
 * This node subscribes to a topic and simply prints out the message contents
 * The node subscribes to the following topics
 * - ~/subs_topic [std_msgs/String]: Subscribing topic
 */

// Include basic header files
#include <ros/ros.h>
// Message header file for subscriber
#include <std_msgs/String.h>

using namespace std;

// Callback (when a message is received) 
void callback_hdlr(const std_msgs::String::ConstPtr& msg, const ros::Subscriber* subs_obj) {
    // Received a message
    ROS_DEBUG_STREAM("Received a message");
    // Info
    ROS_INFO_STREAM("Received message [topic: " << subs_obj->getTopic() 
                    << " (" << subs_obj->getNumPublishers() << ")] '" << msg->data << "'");
}

int main(int argc, char **argv) {
    // Initialize the node
    ros::init(argc, argv, "launch1_subscriber");
    // Create NodeHandle
    ros::NodeHandle nh("~");
    // Create a subscriber object
    ros::Subscriber subs_obj = nh.subscribe<std_msgs::String>("subs_topic", 100, boost::bind(
        callback_hdlr, _1, &subs_obj
    ));
    // Set the node to spin
    ros::spin();
    return 0;
}

File: src/py_basic_nodes/scripts/launch1_subscriber.py
#!/usr/bin/env python3

"""
    This is a simple subscriber that is a part of `launch1`

    This node subscribes to a topic and simply prints out the message contents
    The node subscribes to the following topics
    - subs_topic [std_msgs/String]: Subscribing topic
"""

# Include basic modules
import rospy
import sys
# Modules for messages
from std_msgs.msg import String


# Subscriber callback
def subs_callback(msg: String, subs_obj: rospy.Subscriber):
    # Log
    rospy.logdebug("Received a message")
    rospy.loginfo("Received message [topic: {tname} ({nc})] '{msgdata}'".format(
        tname=subs_obj.name, nc=subs_obj.get_num_connections(),
        msgdata=msg.data
    ))
    pass


# Main function
def main():
    # Initialize node
    rospy.init_node("launch1_subscriber", argv=sys.argv)
    # Create a subscriber
    subs_obj = rospy.Subscriber("subs_topic", String, callback=lambda msg: subs_callback(msg, subs_obj))
    # Set the node to spin indefinitely
    rospy.spin()


# Entry point
if __name__ == '__main__':
    try:
        main()
    except rospy.ROSException as e:
        rospy.logfatal("ROS Exception: {err}".format(err=e))
